Thoughts/example on possible YAML formatting for the protocol file.

jaytmiller opened the pull request:
Assume that anything which can be optional is optional,  e.g. critical_packages.

This is your spec so be brutal if something strikes you the wrong way...  including the whole approach.

jaytmiller committed:
Thoughts/example on possible YAML formatting for the protocol file.

ttdu commented:
I like this. I'm assuming the reason to use YAML instead is because it's easier to parse?

My other thoughts:

- In lieu of (or in addition to) `base_conda_env_spec`, can we offer a `python_version` option? Most curators shouldn't / don't want to specify `gcc` or `pip` versions
- Is is useful to know the list of "primary" packages that define the image. Can we output of the union of `requirements.txt` files _without_ their associated sub-dependencies?
- Curators should not, IMO, be allowed to specify `critical_packages`. I've been debating for a day now about removing this option from the protocol. If it's critical enough to include in the software image, surely we should have a notebook for it, too? I might need to stew on this a bit longer.

jaytmiller commented:
- python_version, sure.   we may want to discuss required tools some but maybe it can be a static list nobody thinks about and just works;  note that some tools need to be installed before pip starts building things
- primary packages, sure.  stage one which involves ingesting the requirements and verifying a dependency solution exists will need some tooling and that can be in there somewhere.  not sure how much tooling,  hopefully not a lot
- critical_packages,  interesting.  (a) I thought curators would define almost all top level dependencies (b) I look at those as the ones people would want to know about in change notes...  although maybe that's the primary packages you mention above.

eteq commented:
I'll weigh in as well on "critical packages" (prompted by @ttdu out-of-band), although with the warning that I am probably missing some context and am happy to be corrected on some incorrect assumption here:

First, I think the name might be misleading.  I might be missing something but from the above it seems to me like "packages that are used in a notebook" are *not* the same as "critical packages", is that right?  If so, that's rather confusing, I would interpret "critical packages" as "packages that are absolutely necessary for the image to work at all" - e.g., something like `jupyterlab/hub`, or even `docker/some container runtime` (although that's implicit in the whole architecture so presumably isn't specified here).  Instead maybe it's some shorter version of "package that is necessary for the science goals of this image"?

Second, I agree with @ttdu in general that if a package is not actually used anywhere, why are we including it?  The  catch is, I'm not sure our current notebook process allows for the possibility of "package that the science community definitely wants, but we don't have the time/expertise to write up a notebook explaining it". So two possible middle-ground solutions:

1. Follow the strict interpretation @ttdu is giving that it must be in a notebook to be included, but make sure we revisit with notebook teams about how to deal with the problem of dependencies that users want but we might not be able to make a *solid* notebook for (the easiest solution to me is to allow either quick-and-dirty notebooks that basically just test imports, or allow external notebooks to be brought in that use the package, but that's in violation of our current quality standards so we'd want some mechanism to flag these to users as something like "shared risk")
2. Allow for this (although with a different name) but enforce that they be "temporary dependencies" - i.e. something like "this dependency was added 10apr2025, so on 11apr2026, it will be auto-removed unless someone has made a notebook by then.  Also the person who added it will get a moderately annoying message that they added a package unnecessarily."  That would ensure we don't fill the images with untested packages in perpetuity, while still allowing short-term "quick fixes"

ttdu commented:
You are correct that `critical packages` is simply a place for curators to say "pretty please I need this really important package that isn't used in a notebook". It could also be used to specify additional version constraints at the image-level. Useful in theory, but I don't think curators should be doing this manually; the notebooks should set the constraints.

Unfortunately, our current process has been riddled with packages that the science community has asked for, but we haven't used in a tutorial (e.g. [Triceratops](https://github.com/stevengiacalone/triceratops), [tess-point](https://github.com/tessgi/tess-point), [tensorflow](https://pypi.org/project/tensorflow/) and other ML packages). We dropped some of these ML packages because it's been so painful building images with them otherwise. This is why I'm so hesitant to allow an exception, even a temporary one, as you outline in 2. I fear it will lead to people requesting temporary patches in perpetuity.

I can support the first solution. I have no objections to pulling in external tutorials, and in fact it's a nice way of politely telling the community that their packages can be installed on TIKE if they write some decent example notebooks. I weep at the thought of relaxing notebook quality standards, but I can accept this as a compromise that allows for emergency deployments of packages to the image.

eteq commented:
So maybe the thing to do is to leave off the critical packages for now, but we should talk in the notebook steering committee about how to implement option 1, @ttdu ?  

If in the end there's some issue with going that road and we decide instead to go with 2, we can add that in later?

jaytmiller commented:
It'll be no problem to add either 1 or 2 later so for now I'll just leave critical_packages out and try to wrap up this first cut.

jaytmiller commented:
In the spirit of keeping things simple,  particularly for the prototype,  I'm also going to **replace** `base_conda_env_spec` with `python_version` but will output a full conda spec detailing tools not added by pip.

jaytmiller committed:
Incorporate spec feedback:
  remove critical_packages
  remove base_conda_env_spec
  add python_version
  promise outputs:
    union of notebook requirements.txt with resolved versions
    full conda spec include Python + build tools
    full pip-tools output detailing versions of all dependencies
